### PR TITLE
Bugfix/save project tree

### PR DESCRIPTION
**Pull request by ichiriac**
fix this issue : https://github.com/HvyIndustries/crane/issues/218#issuecomment-284262836

### CODE REVIEW

File: server/src/util/Storage.ts [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+
+/**
+ * Reads the specified filename
+ */
+export function read(filename: string, cb: (err?: Error, tree?:any) => void): void {
+    if (isProcessing) {
+        // bad things may happen, so delay it by 1 sec
+        setTimeout(() => {
+            read(filename, cb);
+        }, 1000);
```
Comment (nevadascout) [inline server/src/util/Storage.ts]: Is one second too long for this?

Worst case it just delays by another second, so can we shorten this to say, `250ms`?

File: server/src/util/Storage.ts [hunk 2]
```diff
@@ -0,0 +1,136 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Hvy Industries. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *  "HVY", "HVY Industries" and "Hvy Industries" are trading names of JCKD (UK) Ltd
+ *--------------------------------------------------------------------------------------------*/
+
```
Comment (nevadascout) [inline server/src/util/Storage.ts]: Can you refactor this into a class called `Storage` please?

Making this a class would better fit the style of the rest of the server code.

File: server/src/util/Storage.ts (hunk 1 above)
Comment (ichiriac) [inline server/src/util/Storage.ts]: Yes, you are right. I was using 1 sec in order to avoid one edge case, but it's a bad way to handle it.

I think I should use a readFileSync instead, because event if I start before the flag is set, I could potentially read, and write at the same time as actions are async.

Anyway it's a false problem here, as the read function is used at start when the tree is not yet loaded, so there is no way to have a collision in actual system.

File: server/src/util/Storage.ts (hunk 2 above)
Comment (ichiriac) [inline server/src/util/Storage.ts]: Yes of course. I will also expose options as class properties.

**Comment by ichiriac** [comment]
@nevadascout changes done, and also released php-parser with the namespaces prefix bugfix

**Comment by nevadascout** [comment]
Perfect, thanks!
